server/src/main/java/org/apache/iotdb/db/rescon/SystemInfo.java

```diff
@@ -0,0 +1,247 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SystemInfo {
+
+  private static final IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
+  private static final Logger logger = LoggerFactory.getLogger(SystemInfo.class);
+
+  private long totalSgMemCost;
+  private long arrayPoolMemCost;
+  private boolean rejected = false;
+
+  private Map<StorageGroupInfo, Long> reportedSgMemCostMap = new HashMap<>();
+
+  private static final double FLUSH_PROPORTION = config.getFlushProportion();
+  private static final double REJECT_PROPORTION = config.getRejectProportion();
+
+  /**
+   * Report applying a new out of buffered array to system. Attention: It should be invoked before
+   * applying new OOB array actually.
+   *
+   * @param dataType data type of array
+   * @param size     size of array
+   * @return Return true if it's agreed when memory is enough.
+   */
+  public synchronized boolean applyNewOOBArray(TSDataType dataType, int size) {
+    // if current memory is enough
+    if (arrayPoolMemCost + totalSgMemCost + dataType.getDataTypeSize() * size
+        < config.getAllocateMemoryForWrite() * FLUSH_PROPORTION) {
+      arrayPoolMemCost += dataType.getDataTypeSize() * size;
+      logger.debug("Current total mem cost is {}", (arrayPoolMemCost + totalSgMemCost));
+      rejected = false;
+      return true;
+    } else if (arrayPoolMemCost + totalSgMemCost + dataType.getDataTypeSize() * size
+        < config.getAllocateMemoryForWrite() * REJECT_PROPORTION) {
+      arrayPoolMemCost += dataType.getDataTypeSize() * size;
+      // invoke flush()
+      logger.debug("Out of buffer arraies are too large, call for flushing. "
+          + "Current total mem cost is {}", (arrayPoolMemCost + totalSgMemCost));
+      rejected = false;
+      flush();
+      return true;
+    } else {
+      logger.debug("Out of buffer arraies are too large, call for flushing "
+          + "and change system to rejected status...Current total mem cost is {}", 
+          (arrayPoolMemCost + totalSgMemCost));
+      rejected = true;
+      flush();
+      return false;
+    }
+  }
+
+  /**
+   * Report current mem cost of storage group to system.
+   *
+   * @param storageGroupInfo storage group
+   * @param delta mem cost
+   */
+  public synchronized void reportStorageGroupStatus(StorageGroupInfo storageGroupInfo) {
+    long delta = storageGroupInfo.getSgMemCost() - 
+        reportedSgMemCostMap.getOrDefault(storageGroupInfo, 0L);
+    totalSgMemCost += delta;
+    logger.debug("Report Storage Group Status to system. "
+        + "Current array pool mem cost is {}, sg mem cost is {}.", arrayPoolMemCost,
+        totalSgMemCost);
+    reportedSgMemCostMap.put(storageGroupInfo, storageGroupInfo.getSgMemCost());
+
+    long newSgReportThreshold = calculateNewSgReportThreshold(storageGroupInfo);
+    storageGroupInfo.setStorageGroupReportThreshold(newSgReportThreshold);
```

Comment:
We discussed with @jixuan1989 and @qiaojialin, and decided to set a dynamic sgReportThreshold in order to reduce the report frequency. There might be a better way to implement it because I found the current plan is not stable when insert TEXT data.